pkg/sql/catalog/descs/system_descriptors.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package descs
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/config/zonepb"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/bootstrap"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+)
+
+type systemDatabaseNamespaceCache struct {
+	syncutil.RWMutex
+	ns map[descpb.NameInfo]descpb.ID
+}
+
+func (nc *systemDatabaseNamespaceCache) optimisticLookup(
+	parentSchemaID descpb.ID, name string,
+) (wasOptimismWarranted bool, id descpb.ID) {
+	nc.RLock()
+	defer nc.RUnlock()
+	if nc.ns == nil {
+		return false, descpb.InvalidID
+	}
+	return true, nc.ns[descpb.NameInfo{
+		ParentID:       keys.SystemDatabaseID,
+		ParentSchemaID: parentSchemaID,
+		Name:           name,
+	}]
+}
+
+func (nc *systemDatabaseNamespaceCache) populateAll(
+	ctx context.Context, codec keys.SQLCodec,
+) error {
+	nc.Lock()
+	defer nc.Unlock()
```

Comment:
At some point in the future we'll scan the namespace table here for all entries with `parentID = 1`. Right now all system descriptors have hard-coded IDs so there's no point in doing this, we get the IDs from the bootstrapped schema.
Note that we'll have to keep doing this even if we add this namespace table scan: if the cluster is bootstrapped then the namespace table might not have been populated yet, and in any case will be populated with the IDs defined in the bootstrapped schema.